controllers/purge_controller.go

```diff
@@ -0,0 +1,279 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/kyma-project/lifecycle-manager/api/v1beta1"
+	"github.com/kyma-project/lifecycle-manager/pkg/adapter"
+	"github.com/kyma-project/lifecycle-manager/pkg/log"
+	"github.com/kyma-project/lifecycle-manager/pkg/status"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlLog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type RemoteClientResolver func(context.Context, client.ObjectKey) (client.Client, error)
+
+type CRDMatcher func(crd apiextensions.CustomResourceDefinition) bool
+
+// PurgeReconciler reconciles a Kyma object.
+type PurgeReconciler struct {
+	client.Client
+	record.EventRecorder
+	ResolveRemoteClient   RemoteClientResolver
+	PurgeFinalizerTimeout time.Duration
+	SkipCRDs              CRDMatcher
+}
+
+func (r *PurgeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrlLog.FromContext(ctx)
+	logger.V(log.InfoLevel).Info("Purge Reconciliation started")
+
+	ctx = adapter.ContextWithRecorder(ctx, r.EventRecorder)
+
+	// check if kyma resource exists
+	kyma := &v1beta1.Kyma{}
+	if err := r.Get(ctx, req.NamespacedName, kyma); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if kyma.DeletionTimestamp.IsZero() {
+		if err := r.EnsurePurgeFinalizer(ctx, kyma); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// condition to check if deletionTimestamp is set, retry until it gets fully deleted
+	deletionDeadline := kyma.DeletionTimestamp.Add(r.PurgeFinalizerTimeout)
+
+	if time.Now().After(deletionDeadline) {
+		remoteClient, err := r.ResolveRemoteClient(ctx, client.ObjectKeyFromObject(kyma))
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if err := performCleanup(ctx, remoteClient, r.SkipCRDs); err != nil {
+			logger.Error(err, "Finalizer Purging failed")
+			return ctrl.Result{}, err
+		}
+
+		if err := r.DropPurgeFinalizer(ctx, kyma); err != nil {
+			logger.Error(err, "Couldn't remove Purge Finalizer from the Kyma object")
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{
+		RequeueAfter: time.Until(deletionDeadline.Add(time.Second)),
```

Comment:
Why are her 0 Seconds added to the time? If you want to convert it into seconds, you could also cast it. WDYT?

Reply:
`time.Second` resolves to 1 second. So essentially we're just adding a second to the deletion deadline here.